### DISCUSSION TITLE
External Authorization Services

### DISCUSSION THREAD

**Discussion by ctholho** [directus/directus#16090] [Feature Requests]
I think it would be great if you could plug in your own authorization layer instead of solely relying on Directus' permissions. I'm thinking about an external service like [authzed](https://authzed.com), [ory keto](https://github.com/ory/keto), [oso](https://www.osohq.com) or [cerbos](https://cerbos.dev) along with many others (e.g. not open source)

There are good reasons to outsource authentication. Equally, there are good reasons to put authorization into the hands of an external service. With this post, I'm trying to convince you that it should be done and could be done. My question to the maintainers would be, am I way off? While I'm willing to spend a considerable amount of time with the refactor and implementation, I'm not going to fork the project.

## If you're new to authorization
[What is Google Zanzibar](https://authzed.com/blog/what-is-zanzibar/)
[How Carta implemented their own permission system](https://authzed.com/blog/what-is-zanzibar/)
[Oso Authorization Acadamy](https://www.osohq.com/academy)

While Zanzibar seems silly over-engineered for most Directus use-cases. My goal is not to make this into a high availability, global scale powerhouse. It's about the Developer Experience and dev velocity Directus users could gain by outsourcing authorization to an external service.

## What problems does it solve?
* Directus roles and permissions are a sensible start for many projects. Yet, if you want to allow users to create own roles or edit permissions the current implementation makes things hard.
* Auditability: We can easily see permissions by role, but we can't see permissions by object. The above named services usually grant permissions based on ACLs which describe access to an object by the trifecta of **subject, verb, object**. In Directus, the permissions of an object is very easily displayed by roles. If you have few roles and don't use ReBAC you can simply prove who has access. But once you need to use `$CURRENT_USER.tenant` or other relational data it becomes non-trivial.
* Awesome tooling by third parties to write tests for your permissions, visualize access [e.g. authzed playground](https://play.authzed.com/assertions)
* Using Directus with other tools: E.g. while migrating away or to Directus. This would make changing significantly easier because your complex privileges don't need to be reimplemented in another system. Using other tools next to Directus adds the danger of configuration drift.

## Concerns
*  Why use external services (as opposed to implementing proper ABAC/ReBAC ACLs ourselves)?
It's not an easy problem and people have solved it already. Not everybody needs this amount of flexibility. And Directus should focus on its core features.
* I'm not sure 80 % of Directus implementations (blogs or internal asset tools) need this. But you would open Directus to a whole new user segment.
* Why not use hooks?
You can start building apps with the directus-sdk and ory keto *et al* by using hooks. But this way frontend and backend of Directus have to live in separate user spaces.

## What's needed?
An API like we have for endpoints. E.g.: a folder like `extensions/services` containing a file called `authorization.ts` could replace the internal service during at build.

The custom service module needs to offer the same methods. In the case of `AuthorizationServices` it's
* `processAST()`
filter out fields that the user isn't privileged. Only used for `view` permissions.
* `validatePayload()`
verify if the subject is editing or creating objects that adhere to rules, e.g.: Post new transactions only below a value of $1,000,000. Only used for `edit` and `create` actions.
* `checkAccess()`
check if the subject can do `verb` to the object. Used for `read`, `update`, `delete` and `share`.

This is where the community needs to create their own service.

Because the frontend needs to know what a user can do, we need to patch the permissions Services as well. This is where I have the least understanding of the potential complexity. But writing a custom service would entail using the external authorization services method to get more or less all permissions of a user and turning it into something the directus frontend can work with. I think it's not too big of an effort.

I'm interested in your comments.

**Comment by rijkvanzanten** [comment]
I don't think we'll be dropping the "local" authorization any time soon, but it definitely sounds interesting to have a pluggable point in the API where the user can implement their own authorization handler library which can then connect to these third party services. That very closely matches the model we're using for the database and file storage as well! 🙂  Definitely no small feat to build those out though, as permissions checking can get pretty complicated within the system.

I'm interested to see what community engagement this will get. Your explanation makes sense to me, but it's also the first time I've heard it brought up in the years I've been working on this 🤔

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
